### DISCUSSION TITLE
Webhooks support

### DISCUSSION THREAD

**Discussion by qw-in** [SatelCreative/spylib#105] [Ideas]
It would be really helpful if SpyLib had support for webhooks!

Here is my proposal for basic webhook support:

```python
from enum import Enum

from pydantic import BaseModel


class WebhookTopic(Enum):
    ORDERS_CREATE = 'orders/create'
    # ...


class WebhookResponse(BaseModel):
    id: str
    resourceId: str


class Token:
    # ...

    async def webhook_create(
        self,
        topic: WebhookTopic | str,
        callback_url: str,
        include_fields=None,
        metafield_namespaces=None,
        private_metafield_namespaces=None,
    ) -> WebhookResponse:
        """Uses graphql to create a webhook"""
        ...


# Simple but a great building block
# https://shopify.dev/apps/webhooks/configuration/https#step-5-verify-the-webhook -> click python
def verify_webhook(data: str, hmac_header: str, api_secret_key: str) -> bool:
    ...
```

**Comment by hillairet** [comment]
I like the `webhook_create` being part of the Token class since this is one of the operations that one can do with the offline token. Makes sense!
I would add
```python
    async def check_webhook_is_registered(self, topic: WebhookTopic | str, recreate_if_missing: bool = True) -> Optional[WebhookResponse]:
    """Verify if the webhook is still registered and if it isn't, recreate (by default).
    If recreate_if_missing is False this function will return None"""
```
I guess since we return the ids from the create, we could use them as input too ... :thinking: 

For `verify_webhook` we just need this function to call [`utils.hmac.validate`](https://github.com/SatelCreative/spylib/blob/481865b56a73b77b979c6478caf287efc422fd84/spylib/utils/hmac.py#L22) with `is_base64=True`. This argument should be renamed `use_base64` instead of `is_base64` ... that's my bad!  :persevere: 

In the long term it'd be awesome to have something like `check_for_missed_webhooks` that would query the orders within a time range to make sure that no webhook was missed!

**Reply by qw-in** [reply]
```
async def check_webhook_is_registered
```


Would a `webhook_upsert` method work for this purpose?

I'm thinking in that case it can also be used to ensure the args are all consistent

> In the long term it'd be awesome to have something like check_for_missed_webhooks that would query the orders within a time range to make sure that no webhook was missed!

Not sure how we'd achieve that - I'm not aware of a webhook delivery api

**Reply by hillairet** [reply]
I find `upsert` misleading because it doesn't update anything. It checks if the webhook is still all good, which typically it should be, and then does a create if needed. I'd rather have a more meaningful name than a short one.

for the `check_for_missed_webhooks` I meant reading the orders and somehow compared to what was received in the webhooks. We'll need to design this later and separately to figure out how much is done by SPyLib and how much is done by the user, such as how the received webhooks are tracked.

So seems like we have agreement on the first piece which is the `webhook_create` and the `webhook_verify`.

Then we'll do the `check_webhook_is_registered` and finally we can see if/how we would do a webhook missed verification.

Agreed?

**Reply by qw-in** [reply]
> It checks if the webhook is still all good, which typically it should be, and then does a create if needed. I'd rather have a more meaningful name than a short one

It does update things. It would update the following if not set correctly. A more descriptive name is great - I would prefer it starts with `webhook_...` though so it lists alphabetically

```
# These would be updated
callback_url: str,
include_fields=None,
metafield_namespaces=None,
private_metafield_namespaces=None,
```

I agree we can start with the following:
- `webhook_create`
- `webhook_verify`

Then later discuss `check_webhook_is_registered`/`webhook_upsert` and `check_for_missed_webhooks`

**Reply by lishanl** [reply]
🆒

**Comment by qw-in** [comment]
Please note this issue:
- #51
